dotnet/src/SemanticKernel/Kernel.cs

```diff
@@ -194,14 +201,41 @@ public async Task<SKContext> RunAsync(ContextVariables variables, CancellationTo
             try
             {
                 cancellationToken.ThrowIfCancellationRequested();
-                context = await f.InvokeAsync(context, cancellationToken: cancellationToken).ConfigureAwait(false);
+                var functionDetails = f.Describe();
+                string? renderedPrompt = null;
+
+                SemanticFunction? semanticFunction = f as SemanticFunction;
+
+                if (semanticFunction is not null)
+                {
+                    semanticFunction.AddDefaultValues(context.Variables);
+                    renderedPrompt = await semanticFunction._promptTemplate.RenderAsync(context, cancellationToken).ConfigureAwait(false);
+                    context.SetRenderedPrompt(renderedPrompt);
+                }
+
+                if (!this.OnFunctionInvoking(functionDetails, context, renderedPrompt))
+                {
+                    this.Logger.LogInformation("Function invocation was cancelled {0}: {1}.{2}.", pipelineStepCount, f.SkillName, f.Name);
+                    break;
+                }
+
+                if (semanticFunction is not null)
+                {
+                    context = await semanticFunction.InternalInvokeAsync(context, addDefaultValues: false, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
Have you considered updating the signature of InvokeAsync to take a callback?

There's a lot of code paths here, can we add some tests for them?

Reply:
The SetRenderedPrompt on the context feels kind of hacky. What's going to happen if a pre-execution handler ends up calling a semantic function using that context, etc.

Reply:
Thanks for the feedback.

I managed to make it more clean and friendly, moving `renderedPrompt` as a parameter to the `InternalInvokeAsync`. This way I don't need to hacky using the SKContext or using the `bool addDefaultValues` parameter.

Quick note: As the responsibility for rendering prompt is not of the function moving this logic outside of the Invoke implementation is a step forward in that direction as a plan to move the Rendering logic to be a `Plan`.